File: go/behavioral/visitor/visitor_test.go
package visitor

import (
	"testing"
	"../../lib/asserts"
)

func TestVisitor (t *testing.T){

	bladeRunner := NewDvdInfo("Blade Runner", "Harrison Ford", "1")
	electricSheep := NewBookInfo("Do Androids Dream of Electric Sheep?", "Phillip K. Dick")
	sheepRaider := NewGameInfo("Sheep Raider")

	lbv := NewTitleLongBlurbVisitor()
     

	bladeRunner.Accept(lbv)
	asserts.Equals( t,
		"Testing bladeRunner long  " , 
		"LB-DVD: Blade Runner, starring Harrison Ford, encoding region 1",
		lbv.GetTitle())


		electricSheep.Accept(lbv)
		asserts.Equals( t, 	"Testing electricSheep long " , 
			"LB-Book: Do Androids Dream of Electric Sheep?, Author: Phillip K. Dick",
			lbv.GetTitle())

		sheepRaider.Accept(lbv)
		asserts.Equals( t, 	"Testing sheepRaider long " , 
			"LB-Game: Sheep Raider",
			lbv.GetTitle())

	sbv := NewTitleShortBlurbVisitor()

		bladeRunner.Accept(sbv)
		asserts.Equals( t, 	"Testing bladerunner short " , 
			"SB-DVD: Blade Runner",
			sbv.GetTitle())


		electricSheep.Accept(sbv)
		asserts.Equals( t, 	"Testing electricSheep short " ,
			"SB-Book: Do Androids Dream of Electric Sheep?",
			sbv.GetTitle())

		sheepRaider.Accept(sbv)
		asserts.Equals( t, 	"Testing sheepRaider short  " , 
			"SB-Game: Sheep Raider"	,
			sbv.GetTitle())
}      

File: clojure/test/patterns/behavioral/visitor_test.clj
(ns patterns.behavioral.visitor-test
  (:require
      [clojure.test :refer :all]
      [patterns.behavioral.visitor :refer :all]
   ))

(let [
	bladeRunner   (patterns.behavioral.visitor.Dvdinfo.  "Blade Runner" "Harrison Ford"  "1")
	electricSheep (patterns.behavioral.visitor.Bookinfo. "Do Androids Dream of Electric Sheep?" "Phillip K. Dick")
	sheepRaider   (patterns.behavioral.visitor.Gameinfo. "Sheep Raider")
	visitor-long-fn (patterns.behavioral.visitor.Longblurb. )
	visitor-shrt-fn (patterns.behavioral.visitor.Shortblurb.)
	]

(deftest visitor-tst
   (testing "dvd, book, game and long blurb visitor"
		(is (=
			["LB-DVD: Blade Runner, starring Harrison Ford, encoding region: 1"
			"LB-Book: Do Androids Dream of Electric Sheep?, Author: Phillip K. Dick"
			"LB-Game: Sheep Raider"]
			[(accept bladeRunner   visitor-long-fn)
			 (accept electricSheep visitor-long-fn)
			 (accept sheepRaider   visitor-long-fn)] 
		)))

   (testing "dvd, book, game and short blurb visitor"
		(is (=
			["SB-DVD: Blade Runner"
			"SB-Book: Do Androids Dream of Electric Sheep?"
			"SB-Game: Sheep Raider"]
			[(accept bladeRunner   visitor-shrt-fn)
			 (accept electricSheep visitor-shrt-fn)
			 (accept sheepRaider   visitor-shrt-fn)]
		)))
))
